### ISSUE TITLE
loading modules from HTTP sometimes results in multiple downloads

### ISSUE THREAD

**Issue by ry** [denoland/deno#2356]
Here's a minimal example: Run the test http server (`tools/http_server.py`) in the background and call
```
./target/debug/deno run --reload http://localhost:4545/tests/003_relative_import.ts
```
You'll see this:
```
> ./tools/http_server.py
Deno test server http://localhost:4545/
redirect server http://localhost:4546/ -> http://localhost:4545/
redirect server http://localhost:4547/ -> http://localhost:4545/
redirect server http://localhost:4548/ -> http://localhost:4546/
127.0.0.1 - - [14/May/2019 00:08:10] "GET /tests/003_relative_import.ts HTTP/1.1" 200 -
127.0.0.1 - - [14/May/2019 00:08:10] "GET /tests/003_relative_import.ts HTTP/1.1" 200 -
127.0.0.1 - - [14/May/2019 00:08:10] "GET /tests/subdir/print_hello.ts HTTP/1.1" 200 -
127.0.0.1 - - [14/May/2019 00:08:12] "GET /tests/subdir/print_hello.ts HTTP/1.1" 200 -
127.0.0.1 - - [14/May/2019 00:08:12] "GET /tests/subdir/print_hello.ts HTTP/1.1" 200 -
```
Somehow we need to in code this into a Rust unit test. Potentially by using [the Progress bar history](https://github.com/denoland/deno/blob/5e56e26c8baad08954cdc3ecac34923dcdc39c12/cli/progress.rs#L28) (easy) or by mocking out the server (hard)

The actual bug is likely in [`op_fetch_module_meta_data()`](https://github.com/denoland/deno/blob/5e56e26c8baad08954cdc3ecac34923dcdc39c12/cli/ops.rs#L452). It's probably a one line fix. The challenge is to test for this well.

**Comment by kitsonk** [comment]
Could it be a situation where the compiler is requesting a module to have a complete picture of initial module (e.g. TypeScript does a dependency analysis and resolves all modules, requesting their TypeScript source code) before the modules are injected into the runtime and it is then noticed that the module is also needed for that, but it hasn't been transpiled yet, so it goes and fetches it again?

**Comment by ry** [comment]
The compiler thread issues fetchMetaModule ops, which trigger the first set of loads. Because --reload, the cache is bypassed  when V8 loads the compiled ES modules which causes the second load.

**Comment by bartlomieju** [comment]
I've been digging in this issue and I believe I found the culprit. 

When we run:
```
./target/debug/deno run --reload http://localhost:4545/tests/003_relative_import.ts
```
then `DenoFlags.reload` is set to `true` (so don't use cache). 
Then `fetch_module_meta_data_and_maybe_compile_async` call `DenoDir.fetch_module_meta_data_async` which downloads and caches the file. After that since it is TypeScript file `compile_async` is called which spins up the compiler and uses TS compiler to do the job. 

TS compiler the issues `op_fetch_module_meta_data` which calls `DenoDir.fetch_module_meta_data_async` passing the same flags to this method as the first step (so again `DenoFlags.reload` is true) this results in second download because `use_cache` passed to `fetch_module_meta_data_async` is `false`.

We need to figure out how to make compiler download file on first encounter but use cached version of that file for subsequent imports within the same run of program. Hope this helps.

**Comment by bartlomieju** [comment]
Fixed in #2380

**Comment by bartlomieju** [comment]
Fixed in #2477
